File: pkg/galaxy/technologies.go
package galaxy

const ONE_TECH_RESOURCES = 100

type Technologies struct {
	Attack  float64
	Defense float64
	Engine  float64
	Cargo   float64
}

func NewTechnologies() *Technologies {
	return &Technologies{
		Attack:  1.0,
		Defense: 1.0,
		Engine:  1.0,
		Cargo:   1.0,
	}
}

func (t *Technologies) Research(attackResources, defenseResources, engineResources, cargoResources float64) {
	t.Attack += attackResources / ONE_TECH_RESOURCES
	t.Defense += defenseResources / ONE_TECH_RESOURCES
	t.Engine += engineResources / ONE_TECH_RESOURCES
	t.Cargo += cargoResources / ONE_TECH_RESOURCES
}

File: assets/js/entities/technologies.js
export class Technologies {
    /** @type {number} */
    attack = 1;

    /** @type {number} */
    defense = 1;

    /** @type {number} */
    engine = 1;

    /** @type {number} */
    cargo = 1;

    /**
     * @param {Object} data
     * @returns {Technologies}
     */
    updateFromDTO(data) {
        this.attack = data.attack ?? 1;
        this.defense = data.defense ?? 1;
        this.engine = data.engine ?? 1;
        this.cargo = data.cargo ?? 1;
        return this;
    }
}
